File: apps/api/internal/domain/role_discovery.go
package domain

import (
	"encoding/json"
	"time"
)

type RoleSourceKind string

const (
	RoleSourceSearchResult      RoleSourceKind = "Search result"
	RoleSourceManualURL         RoleSourceKind = "Manual URL"
	RoleSourcePastedDescription RoleSourceKind = "Pasted description"
	RoleSourceRecruiter         RoleSourceKind = "Recruiter"
	RoleSourceOther             RoleSourceKind = "Other"
)

type RoleDecisionStatus string

const (
	RoleDecisionNew          RoleDecisionStatus = "New"
	RoleDecisionSaved        RoleDecisionStatus = "Saved"
	RoleDecisionRejected     RoleDecisionStatus = "Rejected"
	RoleDecisionRevisitLater RoleDecisionStatus = "Revisit later"
	RoleDecisionPromoted     RoleDecisionStatus = "Promoted"
)

type RoleFreshnessStatus string

const (
	RoleFreshnessUnknown RoleFreshnessStatus = "Unknown"
	RoleFreshnessLive    RoleFreshnessStatus = "Live"
	RoleFreshnessClosed  RoleFreshnessStatus = "Closed"
)

type RoleRemoteEligibility string

const (
	RoleRemoteUnknown           RoleRemoteEligibility = "Unknown"
	RoleRemoteRemote            RoleRemoteEligibility = "Remote"
	RoleRemoteHybrid            RoleRemoteEligibility = "Hybrid"
	RoleRemoteOnsite            RoleRemoteEligibility = "Onsite"
	RoleRemoteCountryRestricted RoleRemoteEligibility = "Country restricted"
)

type RoleSeniority string

const (
	RoleSeniorityUnknown   RoleSeniority = "Unknown"
	RoleSeniorityJunior    RoleSeniority = "Junior"
	RoleSeniorityMid       RoleSeniority = "Mid"
	RoleSenioritySenior    RoleSeniority = "Senior"
	RoleSeniorityLead      RoleSeniority = "Lead"
	RoleSeniorityStaff     RoleSeniority = "Staff"
	RoleSeniorityPrincipal RoleSeniority = "Principal"
	RoleSeniorityManager   RoleSeniority = "Manager"
)

type RoleCompanyType string

const (
	RoleCompanyUnknown     RoleCompanyType = "Unknown"
	RoleCompanyProduct     RoleCompanyType = "Product"
	RoleCompanyConsultancy RoleCompanyType = "Consultancy"
	RoleCompanyAgency      RoleCompanyType = "Agency"
	RoleCompanyStartup     RoleCompanyType = "Startup"
	RoleCompanyEnterprise  RoleCompanyType = "Enterprise"
	RoleCompanyOther       RoleCompanyType = "Other"
)

type RoleRejectionReason string

const (
	RoleRejectionNone              RoleRejectionReason = ""
	RoleRejectionWrongLocation     RoleRejectionReason = "Wrong location"
	RoleRejectionWrongStack        RoleRejectionReason = "Wrong stack"
	RoleRejectionLowCompensation   RoleRejectionReason = "Low compensation"
	RoleRejectionSeniorityMismatch RoleRejectionReason = "Seniority mismatch"
	RoleRejectionConsultancy       RoleRejectionReason = "Consultancy"
	RoleRejectionDuplicate         RoleRejectionReason = "Duplicate"
	RoleRejectionClosed            RoleRejectionReason = "Closed"
	RoleRejectionOther             RoleRejectionReason = "Other"
)

type RoleSearchTopic struct {
	ID               string
	Name             string
	TargetTitles     string
	PreferredStack   string
	Location         string
	RemotePreference string
	EmploymentType   EmploymentType
	CompanyType      RoleCompanyType
	Compensation     string
	Seniority        RoleSeniority
	Notes            string
	CreatedAt        time.Time
	UpdatedAt        time.Time
}

type RoleRecord struct {
	ID                    string
	SearchTopicID         *string
	Company               string
	Title                 string
	PostingURL            string
	Source                string
	SourceKind            RoleSourceKind
	ProviderSource        string
	Description           string
	RawSourceText         string
	Location              string
	RemoteEligibility     RoleRemoteEligibility
	EmploymentType        EmploymentType
	Seniority             RoleSeniority
	Compensation          string
	Stack                 string
	CompanyType           RoleCompanyType
	FreshnessStatus       RoleFreshnessStatus
	FreshnessCheckedAt    *time.Time
	DecisionStatus        RoleDecisionStatus
	RejectionReason       RoleRejectionReason
	PromotedApplicationID *string
	Metadata              json.RawMessage
	CreatedAt             time.Time
	UpdatedAt             time.Time
}

type RoleSearchRequest struct {
	Topic    RoleSearchTopic
	MaxRoles int
}

type RoleSearchProviderResult struct {
	Company            string
	Title              string
	PostingURL         string
	Source             string
	ProviderSource     string
	Description        string
	RawSourceText      string
	Location           string
	RemoteEligibility  RoleRemoteEligibility
	EmploymentType     EmploymentType
	Seniority          RoleSeniority
	Compensation       string
	Stack              string
	CompanyType        RoleCompanyType
	FreshnessStatus    RoleFreshnessStatus
	FreshnessCheckedAt *time.Time
	Metadata           json.RawMessage
}

type ImportedRoleSummary struct {
	RoleID     string
	Company    string
	Title      string
	PostingURL string
}

type SkippedRoleSummary struct {
	Company    string
	Title      string
	PostingURL string
	Reason     string
}

type RoleSearchRunResult struct {
	TopicID       string
	ImportedCount int
	SkippedCount  int
	Imported      []ImportedRoleSummary
	Skipped       []SkippedRoleSummary
}

func ValidateRoleDecisionStatus(status RoleDecisionStatus) error {
	switch status {
	case RoleDecisionNew, RoleDecisionSaved, RoleDecisionRejected, RoleDecisionRevisitLater, RoleDecisionPromoted:
		return nil
	default:
		return ErrInvalidRoleDecisionStatus
	}
}

func ValidateRoleFreshnessStatus(status RoleFreshnessStatus) error {
	switch status {
	case RoleFreshnessUnknown, RoleFreshnessLive, RoleFreshnessClosed:
		return nil
	default:
		return ErrInvalidRoleFreshnessStatus
	}
}

File: apps/web/src/domain/roleDiscovery.ts
import type { EmploymentType } from "./jobOpportunity";

export const roleDecisionStatuses = [
  "New",
  "Saved",
  "Rejected",
  "Revisit later",
  "Promoted"
] as const;

export type RoleDecisionStatus = (typeof roleDecisionStatuses)[number];

export const roleFreshnessStatuses = ["Unknown", "Live", "Closed"] as const;

export type RoleFreshnessStatus = (typeof roleFreshnessStatuses)[number];

export const roleRemoteEligibilities = [
  "Unknown",
  "Remote",
  "Hybrid",
  "Onsite",
  "Country restricted"
] as const;

export type RoleRemoteEligibility = (typeof roleRemoteEligibilities)[number];

export const roleSourceKinds = [
  "Search result",
  "Manual URL",
  "Pasted description",
  "Recruiter",
  "Other"
] as const;

export type RoleSourceKind = (typeof roleSourceKinds)[number];

export const roleSeniorities = [
  "Unknown",
  "Junior",
  "Mid",
  "Senior",
  "Lead",
  "Staff",
  "Principal",
  "Manager"
] as const;

export type RoleSeniority = (typeof roleSeniorities)[number];

export const roleCompanyTypes = [
  "Unknown",
  "Product",
  "Consultancy",
  "Agency",
  "Startup",
  "Enterprise",
  "Other"
] as const;

export type RoleCompanyType = (typeof roleCompanyTypes)[number];

export const roleRejectionReasons = [
  "",
  "Wrong location",
  "Wrong stack",
  "Low compensation",
  "Seniority mismatch",
  "Consultancy",
  "Duplicate",
  "Closed",
  "Other"
] as const;

export type RoleRejectionReason = (typeof roleRejectionReasons)[number];

export type RoleSearchTopic = {
  id: string;
  name: string;
  targetTitles: string;
  preferredStack: string;
  location: string;
  remotePreference: string;
  employmentType: EmploymentType;
  companyType: RoleCompanyType;
  compensation: string;
  seniority: RoleSeniority;
  notes: string;
  createdAt: string;
  updatedAt: string;
};

export type RoleSearchTopicCommand = Omit<
  RoleSearchTopic,
  "id" | "createdAt" | "updatedAt"
>;

export type RoleRecord = {
  id: string;
  searchTopicId: string | null;
  company: string;
  title: string;
  postingUrl: string;
  source: string;
  sourceKind: RoleSourceKind;
  providerSource: string;
  description: string;
  rawSourceText: string;
  location: string;
  remoteEligibility: RoleRemoteEligibility;
  employmentType: EmploymentType;
  seniority: RoleSeniority;
  compensation: string;
  stack: string;
  companyType: RoleCompanyType;
  freshnessStatus: RoleFreshnessStatus;
  freshnessCheckedAt: string | null;
  decisionStatus: RoleDecisionStatus;
  rejectionReason: RoleRejectionReason;
  promotedApplicationId: string | null;
  metadata: string;
  createdAt: string;
  updatedAt: string;
};

export type RoleRecordCommand = Omit<
  RoleRecord,
  | "id"
  | "freshnessCheckedAt"
  | "decisionStatus"
  | "rejectionReason"
  | "promotedApplicationId"
  | "createdAt"
  | "updatedAt"
>;

export type RoleRecordsFilter = {
  decisionStatus?: RoleDecisionStatus;
  freshnessStatus?: RoleFreshnessStatus;
  sourceKind?: RoleSourceKind;
  searchTerm?: string;
};

export type ImportedRoleSummary = {
  roleId: string;
  company: string;
  title: string;
  postingUrl: string;
};

export type SkippedRoleSummary = {
  company: string;
  title: string;
  postingUrl: string;
  reason: string;
};

export type RoleSearchRunResult = {
  topicId: string;
  importedCount: number;
  skippedCount: number;
  imported: ImportedRoleSummary[];
  skipped: SkippedRoleSummary[];
};

export type PromoteRoleResult = {
  role: RoleRecord;
  applicationId: string;
};
